### PR TITLE
Add support to PropertyDataFetcher for aliased props from a Map Datasource

### PR DESCRIPTION
**Pull request by swaters-atlassian**
When the data source for the PropertyDataFetcher is a Map the keys will either be the property name or the aliased property name if the query specified an alias. Currently the PropertyDataFetcher only attempts to lookup the value by the property name which means aliased values will resolve as null.
This change checks if a field has been aliased when doing a Map lookup and attempts to retrieve the value with the alias vs the property name.

In this query case, the map for the PropertyDataFetcher will be mapped by the alias (nameAlias) not name.
```graphql
query($id: String!) {
  foo(id: $id) {
    id
    nameAlias: name
  }
}
```

I'm not sure if I should have the logic fall back to fetching by the propertyName if no value is found for the alias, this seems like it would be reasonable, but I don't have enough context to be sure if that's the correct behavior.

### CODE REVIEW

**Comment by andimarek** [comment]
to give a bit more context about the general problem:
(it was also raised before in https://github.com/graphql-java/graphql-java/issues/826)
the problem is how aliases are mapped to java objects/maps.

For example:
```graphql
foo {
  id
  alias1: name
  alias2: name(notDefault:true)
}
```

If a DataFetcher for `foo` wants to return the fully resolved data for the whole sub-query (not only one value for `foo` which will be then used as source for `id`, and `name` (two times)) how can that be achieved?

I am not really sure how special this requirement is and if we should make support it out of the box. 

Another thought: Maybe this is also not really a DataFetcher issue at all and we wanna be able to return the already resolved sub-query and make the execution mark as done for that sub-query somehow? @bbakerman

**Comment by bbakerman** [comment]
I always took "alias" to be a renaming technique, not how you should get to the underlying data.

That is I have a field called  `name` BUT the client making the query wants to call it `aliasedName` on this query.  So its very much driven by the client making the query, not the server providing the schema.

in graphql now we have the @fetch aliasing that allows the server (not the client) to indicate that property name should be used

    type Query {
        name : String @fetch(from:"anotherProperty")
    }

My other problem with this PR is that is handles maps but not POJO methods.  It should be consistent on all access strateegies

In short I dont think we should do this.  aliasing is a client side thing to get the same property I think

**Comment by swaters-atlassian** [comment]
Interestingly, the existing PropertyDataFetcher does actually handle aliases properly in POJOs which is to some degree why I chose to fix things in this particular way.  The reason it works for POJOs is that the lookup is by the getter of the propertyName on the object which will work regardless of the alias.  It fails in the alias case where a map is supplied because in order to support cases like returning the same property twice with different formats it makes a certain amount of sense to key the map by the alias rather than the propertyName so both can exist.  However, to resolve that requires querying the map by the alias and not the propertyName if an alias exists.

If there is a different place other than the DataFetcher where we should be resolving aliases then help me out.  I can see a couple alternatives for our use case:
1. The source map is built up with only propertNames as keys and aliased properties (and their associated arguments) are mapped in and resolved somewhere higher up in the chain after fetching.
2. We write our own DataFetcher and attempt to patch it in as the default resolver for our needs

**Comment by andimarek** [comment]
fyi: this (or other solutions) will not go into 9.0

**Comment by andimarek** [comment]
I will close that PR because this is a problem on the execution level: how can you delegate larger parts of the executions to something else, not just a DataFetcher.

A DataFetcher is just responsible for getting the data for the field based on the `source` context. Taking the alias into account actually happens on the execution level.
